app/services/oneoffs/populate_missing_trns.rb

```diff
@@ -0,0 +1,80 @@
+# frozen_string_literal: true
+
+require "has_recordable_information"
+
+module Oneoffs
+  class PopulateMissingTrns
+    include HasRecordableInformation
+
+    def perform_change(dry_run: true)
+      reset_recorded_info
+
+      record_info("~~~ DRY RUN ~~~") if dry_run
+
+      ActiveRecord::Base.transaction do
+        teacher_profiles_without_trns.find_each do |teacher_profile|
+          trns = lookup_trns(teacher_profile)
+          next if trns.empty?
+
+          teacher_profile.update!(trn: trns.first)
+
+          record_info("teacher profile TRN updated to #{teacher_profile.trn} for teacher profile #{teacher_profile.id}")
+          record_info("multiple TRNs found for teacher profile #{teacher_profile.id}: #{trns.join}") if trns.count > 1
+        end
+
+        raise ActiveRecord::Rollback if dry_run
+      end
+
+      recorded_info
+    end
+
+  private
+
+    def lookup_trns(teacher_profile)
+      users = associated_users(teacher_profile.user_id)
+
+      trns_from_npq_applications = lookup_trns_from_npq_appliations(users)
+      trns_from_ecf_validation = lookup_trns_from_ecf_validation(users)
+      trns_from_teacher_profiles = lookup_trns_from_teacher_profiles(users)
+
+      (trns_from_npq_applications + trns_from_ecf_validation + trns_from_teacher_profiles).uniq.select do |trn|
+        TeacherReferenceNumber.new(trn).valid?
+      end
+    end
+
+    def lookup_trns_from_npq_appliations(users)
+      users.map { |user|
+        user
+          .npq_applications
+          .where(teacher_reference_number_verified: true)
+          .pluck(:teacher_reference_number)
+      }.flatten
+    end
+
+    def lookup_trns_from_ecf_validation(users)
+      participant_profiles = users.map(&:participant_profiles).flatten
+
+      participant_profiles.map { |participant_profile|
+        participant_profile.ecf_participant_validation_data&.trn
```

Comment:
Just to confirm this record ensures that the ten is verified?

Reply:
Good point, I was checking for NPQ applications but not the others, which seem to be inferred from the `ECFParticipantEligibility`. I've added that in now.